apps/astarte_housekeeping_api/lib/astarte_housekeeping_api/realms/realm.ex

```diff
@@ -93,15 +114,32 @@ defmodule Astarte.Housekeeping.API.Realms.Realm do
           "must be used with replication_class NetworkTopologyStrategy"
         )
 
-      get_field(changeset, :replication_factor) == nil ->
-        changeset
-        |> put_change(:replication_factor, @default_replication_factor)
-
       true ->
         changeset
     end
   end
 
+  defp maybe_put_default_replication_factor(changeset) do
+    replication_class = get_field(changeset, :replication_class)
+    replication_factor = get_field(changeset, :replication_factor)
+
+    if replication_class == @default_replication_class and replication_factor == nil do
+      put_change(changeset, :replication_factor, @default_replication_factor)
+    else
+      changeset
+    end
+  end
+
+  defp validate_pem_public_key_if_present(changeset, field) do
```

Comment:
I don't think this is correct. The pem will always be present because it is required during the creation, and so will always be `!= nil` (unless manually set to `nil` in `params`).

What you probably want is to match `with {:changes, new_pem} <- fetch_field(changeset, field)`, so that it will only validate changes.

BUT there already is a function which only validates if we're editing the value, and that is `Ecto.Changeset.validate_change/3`.

So I suggest instead rewriting `validate_pem_public_key/2` as 

```elixir
defp validate_pem_public_key(:jwt_public_key_pem, pem) do
  try do
    [{:SubjectPublicKeyInfo, _, _}] = :public_key.pem_decode(pem)
    []
  rescue
    _ ->
      [jwt_public_key_pem: "is not a valid PEM public key"]
  end
end
```

and using it with `validate_change/2`. This would also work for `changeset` as it should be used as a creation changeset on an empty struct.

Also, even in `update_changeset` we want `:jwt_public_key_pem` required, because it will only invalidate the changeset if the value isn't present after the change.

```elixir
iex(15)> Ecto.Changeset.change(%HK{realm_name: "test"}, %{}) |> Ecto.Changeset.validate_required(:jwt_public_key_pem)
#Ecto.Changeset<
  action: nil,
  changes: %{},
  errors: [jwt_public_key_pem: {"can't be blank", [validation: :required]}],
  data: #Astarte.Housekeeping<>,
  valid?: false
>
iex(16)> Ecto.Changeset.change(%HK{realm_name: "test"}, %{jwt_public_key_pem: "pem"}) |> Ecto.Changeset.validate_required(:jwt_public_key_pem)
#Ecto.Changeset<
  action: nil,
  changes: %{jwt_public_key_pem: "pem"},
  errors: [],
  data: #Astarte.Housekeeping<>,
  valid?: true
>
iex(17)> Ecto.Changeset.change(%HK{realm_name: "test", jwt_public_key_pem: "pem"}, %{}) |> Ecto.Changeset.validate_required(:jwt_public_key_pem)
#Ecto.Changeset<action: nil, changes: %{}, errors: [],
 data: #Astarte.Housekeeping<>, valid?: true>
iex(18)> Ecto.Changeset.change(%HK{realm_name: "test", jwt_public_key_pem: "pem"}, %{jwt_public_key_pem: nil}) |> Ecto.Changeset.validate_required(:jwt_public_key_pem)
#Ecto.Changeset<
  action: nil,
  changes: %{},
  errors: [jwt_public_key_pem: {"can't be blank", [validation: :required]}],
  data: #Astarte.Housekeeping<>,
  valid?: false
>
```

so we should still keep `validate_required(@required_fields)`

Reply:
What you want to validate here is not a full realm configuration, but just a merge patch (see https://datatracker.ietf.org/doc/html/rfc7386) to the realm in order to update it. Therefore, the public key (as any other field) might not be present. As an example, this is a valid (albeit useless) patch: `{"data": {}}`.
The backend will then take care of updating the realm, merging the patch and the current status, and for that reason it will return an updated realm.